Fb 221 i18n.t to hook to contentful

bhavanaananda opened the pull request:
Issue: https://dfedigital.atlassian.net/browse/FB-221

- Added contentful_migrate to migrate from en.yml to Contentful- 'translation'
- Added a custom I18n backend - Contentful for ENV:'production' and fallback as yml
- ENV:'development' currently works on fallback yml

![Screenshot 2025-07-07 at 12 53 48](https://github.com/user-attachments/assets/b34a31b3-b431-464f-b92a-f0b13068651f)

**‘Translations’ in  Contentful - Local Contentful instance**

![image](https://github.com/user-attachments/assets/cd949a8b-909d-45ec-af31-a7ba0d6290af)

![image](https://github.com/user-attachments/assets/bd66d8d6-8e76-469d-887a-1fcf48671179)

bhavanaananda commented on config/environments/production.rb:
```diff
@@ -43,8 +44,14 @@
-  # config.cache_store = :mem_cache_store
+  config.cache_store = :redis_cache_store,
+                       {
+                         url: ENV["REDIS_URL"],
```
I have added fetch

### For Production:

RAILS_ENV = production
 find-a-buying-solution % REDIS_URL=redis://localhost:6379 RAILS_ENV=production bundle exec rails assets:precompile
**I18n: Using Contentful backend with YAML fallback**
**I18n backends initialized: I18n::Backend::Contentful, I18n::Backend::Simple**
its loading contentful 

Heroku already has REDIS_URL set up in its Config vars

But to simulate production locally I installed local redis and ran on a diff terminal

Then ran :
RAILS_ENV=production bundle exec rails assets:precompile
I18n: Using Contentful backend with YAML fallback
I18n backends initialized: I18n::Backend::Contentful, I18n::Backend::Simple
.....

Accessed 
http://localhost:3000/
http://localhost:3000/solutions

All load from cached contentful-"translations' hash and seamlessly


### For Development:

Its falls back to YAML

--------------
### REDIS-SERVER
To simulate production locally I installed local redis and ran on a diff terminal
You don't need to do this for ENV=development as it falls back to yml

 find-a-buying-solution % redis-server
35543:C 08 Jul 2025 15:25:40.202 * oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
35543:C 08 Jul 2025 15:25:40.202 * Redis version=8.0.3, bits=64, commit=00000000, modified=1, pid=35543, just started
35543:C 08 Jul 2025 15:25:40.202 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
35543:M 08 Jul 2025 15:25:40.202 * monotonic clock: POSIX clock_gettime
                _._                                                  
           _.-``__ ''-._                                             
      _.-``    `.  `_.  ''-._           Redis Open Source            
  .-`` .-```.  ```\/    _.,_ ''-._      8.0.3 (00000000/1) 64 bit
 (    '      ,       .-`  | `,    )     Running in standalone mode
 |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
 |    `-._   `._    /     _.-'    |     PID: 35543
  `-._    `-._  `-./  _.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |           https://redis.io       
  `-._    `-._`-.__.-'_.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |                                  
  `-._    `-._`-.__.-'_.-'    _.-'                                   
      `-._    `-.__.-'    _.-'                                       
          `-._        _.-'                                           
              `-.__.-'                                               

35543:M 08 Jul 2025 15:25:40.203 # WARNING: The TCP backlog setting of 511 cannot be enforced because kern.ipc.somaxconn is set to the lower value of 128.
35543:M 08 Jul 2025 15:25:40.203 * Server initialized
35543:M 08 Jul 2025 15:25:40.204 * Loading RDB produced by version 8.0.3
35543:M 08 Jul 2025 15:25:40.204 * RDB age 16 seconds
35543:M 08 Jul 2025 15:25:40.204 * RDB memory usage when created 0.76 Mb
35543:M 08 Jul 2025 15:25:40.204 * Done loading RDB, keys loaded: 0, keys expired: 0.
35543:M 08 Jul 2025 15:25:40.204 * DB loaded from disk: 0.000 seconds
35543:M 08 Jul 2025 15:25:40.204 * Ready to accept connections tcp

bhavanaananda replied:
> this should probably use `fetch` to fail early on misconfiguration.

done.

threedaymonk commented on lib/tasks/contentful_migrate.rake:
```diff
@@ -0,0 +1,81 @@
+      else
+        puts "Failed to create the '#{key}': #{response.body}"
+      end
+      # Adding tiny delay between requests to account for race conditions
+      sleep(0.5)
```
what's the race condition here?

threedaymonk commented on lib/i18n/backend/contentful.rb:
```diff
@@ -0,0 +1,104 @@
+            end
+          end
+
+          Rails.cache.write("contentful_translations", cached_translations, expires_in: 1.hour)
```
Can these two magic values be named constants?

bhavanaananda replied:
done.

bhavanaananda commented on lib/i18n/backend/contentful.rb:
```diff
@@ -0,0 +1,104 @@
+            limit: 1000
+          )
+
+          cached_translations = entries.each_with_object({}) do |entry, hash|
+            next if entry.fields[:key].blank?
+
+            key_parts = entry.fields[:key].split(".")
+            next if key_parts.empty?
+
+            locale = key_parts.shift.to_sym
+
+            hash[locale] ||= {}
+            current = hash[locale]
+
+            key_parts[..-2].each do |part|
+              next if part.blank?
+
+              current[part.to_sym] ||= {}
+              current = current[part.to_sym]
+            end
+
+            if key_parts.last.present?
+              current[key_parts.last.to_sym] = entry.fields[:value]
+            end
+          end
```
Yes those translate the flattened migrated values in contentful into hash and the last value in it carries the value.

bhavanaananda replied:
Done.

bhavanaananda commented on lib/i18n/backend/contentful.rb:
```diff
@@ -0,0 +1,104 @@
+        end
+
+        flat_key = split_keys[1..].join(".")
+        val = translations[split_keys[0]]&.[](flat_key.to_sym)
```
done.

bhavanaananda commented on lib/i18n/backend/contentful.rb:
```diff
@@ -0,0 +1,104 @@
+
+      def initialized?
+        !@translations.nil?
+      end
```
removed, done.

threedaymonk commented on config/environments/production.rb:
```diff
@@ -43,8 +44,12 @@
 
-  # Replace the default in-process memory cache store with a durable alternative.
-  # config.cache_store = :mem_cache_store
+  config.cache_store = :redis_cache_store,
+                       {
+                         url: ENV.fetch("REDIS_URL"),
+                         ssl_params: { verify_mode: OpenSSL::SSL::VERIFY_NONE },
+                         expires_in: 24.hours,
+                       }
```
In any other context, I'd think assignment with a comma was a mistake, but this is indeed the recommended way of setting it up!

bhavanaananda commented on lib/i18n/utils.rb:
```diff
@@ -0,0 +1,44 @@
+module I18n
+  module Utils
```
Done.

threedaymonk reviewed (changes requested):
This is excellent work!

For our own convenience I think we also need some rake tasks to help synchronise Contentful and `en.yml`, but you might want to put those in a separate PR:

1. Upload any new translations we've added to `en.yml`
2. List and/or delete any translations that are in Contentful but not in `en.yml` (to find ones we've deleted)
3. Update `en.yml` with any translations that have changed in Contentful.

bhavanaananda commented:
> This is excellent work!
> 
> For our own convenience I think we also need some rake tasks to help synchronise Contentful and `en.yml`, but you might want to put those in a separate PR:
> 
> 1. Upload any new translations we've added to `en.yml` to Contentful
> 2. List and/or delete any translations that are in Contentful but not in `en.yml` (to find ones we've deleted)
> 3. Update `en.yml` with any translations that have changed in Contentful.

@threedaymonk shall I create a new ticket for this?

bhavanaananda commented on config/initializers/i18n.rb:
```diff
@@ -0,0 +1,57 @@
+                      Rails.logger.warn "Failed to initialize Contentful I18n backend: #{e.message}. Falling back to YAML translations."
+                      [yaml_backend]
+                    end
+                  end
```
Done. Can you take a look at the changes made for development.rb and production.rb? 
config/initializers/i18n.rb is now deleted as its redundant.
